src/Microsoft.AspNetCore.Razor.Evolution/ITagHelperBoundAttributeDescriptorBuilder.cs

```diff
@@ -0,0 +1,282 @@
+                        _containingTypeName,
+                        _propertyName,
+                        _indexerNamePrefix);
+
+                    yield return diagnostic;
+                }
+                else if (_indexerNamePrefix.Length > 0 && string.IsNullOrWhiteSpace(_indexerNamePrefix))
+                {
+                    var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundAttributeNullOrWhitespace(
+                        _containingTypeName,
+                        _propertyName);
+
+                    yield return diagnostic;
+                }
+                else
+                {
+                    foreach (var character in _indexerNamePrefix)
+                    {
+                        if (char.IsWhiteSpace(character) || InvalidNonWhitespaceAttributeNameCharacters.Contains(character))
+                        {
+                            var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundAttributePrefix(
+                                _containingTypeName,
+                                _propertyName,
+                                _indexerNamePrefix,
+                                character);
+
+                            yield return diagnostic;
+                        }
+                    }
+                }
+            }
+        }
+
+        private void EnsureDiagnostics()
+        {
+            if (_diagnostics == null)
+            {
+                _diagnostics = new HashSet<RazorDiagnostic>();
+            }
+        }
+
+        private class ITagHelperBoundAttributeDescriptor : BoundAttributeDescriptor
+        {
+            public ITagHelperBoundAttributeDescriptor(
+                bool isEnum,
+                string name,
+                string propertyName,
+                string typeName,
+                string dictionaryAttributeNamePrefix,
+                string dictionaryValueTypeName,
+                string documentation,
+                string displayName,
+                Dictionary<string, string> metadata,
+                IReadOnlyList<RazorDiagnostic> diagnostics) : base(DescriptorKind)
+            {
+                IsEnum = isEnum;
+                IsIndexerStringProperty = dictionaryValueTypeName == typeof(string).FullName || dictionaryValueTypeName == "string";
+                IsStringProperty = typeName == typeof(string).FullName || typeName == "string";
+                Name = name;
+                TypeName = typeName;
+                IndexerNamePrefix = dictionaryAttributeNamePrefix;
+                IndexerTypeName = dictionaryValueTypeName;
+                Documentation = documentation;
+                DisplayName = displayName;
+                Diagnostics = diagnostics;
+
+                metadata[PropertyNameKey] = propertyName;
+                Metadata = new Dictionary<string, string>(metadata);
```

Comment:
you should not dirty your parameter (see L277), make a copy first and then modify the copy.